### PR TITLE
Add toEVMInstruction helper

### PR DESCRIPTION
**Pull request by axic**
Adds https://github.com/ethereum/solidity/pull/11192#discussion_r604457094

Not sure it is worth it though.

### CODE REVIEW

File: libyul/optimiser/OptimizerUtilities.cpp [hunk 1]
```diff
@@ -47,3 +49,12 @@ bool yul::isRestrictedIdentifier(Dialect const& _dialect, YulString const& _iden
 {
 	return _identifier.empty() || TokenTraits::isYulKeyword(_identifier.str()) || _dialect.reservedIdentifier(_identifier);
 }
+
+optional<evmasm::Instruction> yul::toEVMInstruction(Dialect const& _dialect, YulString const& _name)
+{
+	if (auto dialect = dynamic_cast<EVMDialect const*>(&_dialect))
+		if (auto builtin = dialect->builtin(_name))
```
Comment (cameel) [inline libyul/optimiser/OptimizerUtilities.cpp]:
```suggestion
	if (auto const* dialect = dynamic_cast<EVMDialect const*>(&_dialect))
		if (BuiltinFunctionForEVM const* builtin = dialect->builtin(_name))
```

File: libyul/optimiser/OptimizerUtilities.cpp [hunk 2]
```diff
@@ -47,3 +49,12 @@ bool yul::isRestrictedIdentifier(Dialect const& _dialect, YulString const& _iden
...
+			if (builtin->instruction)
+				return *builtin->instruction;
```
Comment (cameel) [inline libyul/optimiser/OptimizerUtilities.cpp]: Looks like `instruction` is an `optional` already. No point in unpacking and packing it again:
```suggestion
			return builtin->instruction;
```

**Review by cameel — DISMISSED** [review]
I think it's worth it. Does not save all that much code but wraps it in a nice name which does improve readability, at least for me.

I think it's fine already but I have two suggestions which would make it slightly better.

File: libyul/optimiser/OptimizerUtilities.cpp (hunk 2 above)
Comment (axic) [reply libyul/optimiser/OptimizerUtilities.cpp]: Good catch!

**Review by cameel — APPROVED** [review]

**Review by hrkrshnn — APPROVED** [review]

File: libyul/optimiser/Metrics.cpp [hunk 3]
```diff
@@ -139,13 +139,11 @@ void CodeCost::operator()(FunctionCall const& _funCall)
...
 	ASTWalker::operator()(_funCall);
 
-	if (EVMDialect const* dialect = dynamic_cast<EVMDialect const*>(&m_dialect))
-		if (BuiltinFunctionForEVM const* f = dialect->builtin(_funCall.functionName.name))
-			if (f->instruction)
-			{
-				addInstructionCost(*f->instruction);
-				return;
-			}
+	if (auto instruction = toEVMInstruction(m_dialect, _funCall.functionName.name))
```
Comment (chriseth) [inline libyul/optimiser/Metrics.cpp:142]: Can we please agree not to use `auto` excessively?

Comment (axic) [reply libyul/optimiser/Metrics.cpp:142]: Even though this line wasn't, the other places were using auto prior, hence that's why I kept using it.

**Pull request merged**